### ISSUE TITLE
Colors are not real after applying correction

### ISSUE THREAD

**Issue by alihamidisepehr** [danforthcenter/plantcv#828]
**Describe the bug**
I'm using [this tutorial](https://plantcv.readthedocs.io/en/v3.11.0/transform_color_correction_tutorial/) to correct colors in photos from melons (Scenario B). Since colorchecker was tilted in some of my images, I tried to manually check  if the mask is located in color chips. I followed all the steps but images after correction were either too bright or in some cases too dark. I also used the regression plot you demonstrated at the end of the tutorial and I also got a very high correlation.

**To Reproduce*
![IMG_2153](https://user-images.githubusercontent.com/43645711/139707289-4bb10875-36df-44b2-852e-3ee6b1fef131.JPG)
![IMG_2176](https://user-images.githubusercontent.com/43645711/139707303-91bc4712-ba9a-4de6-b6a5-ca820f49456e.JPG)
![image](https://user-images.githubusercontent.com/43645711/139707928-7383d62f-f7d0-4f35-a74a-ada6104f4ce4.png)

I was expecting each color chip would match between target and source (blue to blue, red to red) after correction but I don't think that's what's happening here. Could you please help me with this?

Also, can you please explain what are the recommended criteria for a target image? Do I have to have a perfect image with only colorchecker in it? if so, can I crop one of my photos to generate a target image?

Thank you in advance!

**Comment by alihamidisepehr** [comment]
Hi @nfahlgren, I was wondering if I should wait for any response here or there's going to be an update in the code? Thank you

**Comment by nfahlgren** [comment]
Hi @alihamidisepehr,

Sorry, this was automatically closed when we merged the pull request with the bugfix. We have had a few people report issues with color correction with similar issues to what you reported above, thank you! @JorgeGtz identified a bug in the module and developed the fix for it in #829. He tested it with your images and it looks like it corrects the issue.

I created a bugfix release (v3.13.1) that includes the updates. It's available on PyPI already and can be installed via `pip` and should be available through `conda` in a few hours.

**Comment by alihamidisepehr** [comment]
Thank you @nfahlgren. I'm looking forward to testing it with new edits. In my images I have images with different lighting conditions and in some of them there's also shadows caused by the colorspace on the lower surfaces including the side of the melon. Do you recommend using LAB colorspace or I should stay with RGB?

Thank you,
Ali

**Comment by alihamidisepehr** [comment]
@nfahlgren @JorgeGtz I did `Pip install plantcv` again. I also have another concern. My understanding about the color correction with this approach is to have colorchecker with same rgb values for target image and a corrected image (brown in target and corrected source should be same or at least closer compared to the source image before correction). Is that right?
If so, I'm having different numbers for example from brown chip in the target image and in the corrected image. I have many source images with a colorchecker in each and I'm trying to correct colors on all of them consistently. Please advise. Appreciate your support.

**Comment by nfahlgren** [comment]
Correct, if you used one image as the target to correct many source images then you would expect the chip color values in the corrected images to be similar to the chip colors in the uncorrected target image. `pcv.transform.quick_color_check` can be used to plot a comparison of the target and source color cards: https://plantcv.readthedocs.io/en/stable/quick_color_check/

**Comment by alihamidisepehr** [comment]
Thank you for your prompt response. Yes, I used that too. It shows a very high correlation.
![image](https://user-images.githubusercontent.com/43645711/140421155-4f73c292-0fc5-4145-8d68-235539b2265a.png)
To be able to reproduce it, here's an example of source and target:
Target
![IMG_2176](https://user-images.githubusercontent.com/43645711/140421300-d3e81455-118e-4f8d-addb-dac54a5885c1.JPG)

Source
![IMG_2198](https://user-images.githubusercontent.com/43645711/140421385-d7fb07fb-6889-4167-8cd8-ce63930b3d04.JPG)
I also used GIMP to check RGB values on each chip and I saw a big difference. The corrected source for me is too bright. This is what I get
![image](https://user-images.githubusercontent.com/43645711/140421681-83c36ccc-4321-40b7-b31a-e3df02609da7.png)
Not sure what I should do differently. Thank you

**Comment by JorgeGtz** [comment]
Hi @alihamidisepehr, I tried to replicate the issue and this is what I get
![melon_source_corrected](https://user-images.githubusercontent.com/26798565/140425057-92c988d8-f3ff-4e33-bec0-e310899c4807.jpg)
I think some chips get brighter than the target, but not as much as in your output.
Could you run `pcv.__version__`  to verify that you got version 3.13.1 installed?

**Comment by alihamidisepehr** [comment]
@JorgeGtz @nfahlgren thank you for your assistance. Yes, you're right. For some reason I was still using 3.13.0. I could replicate the same result as you. Just a couple of other points:
when I tried `pcv.transform.quick_color_check`, I got this result with only one data point:
![image](https://user-images.githubusercontent.com/43645711/140524501-9c102caa-4fbc-443d-86d0-7567ec23c69f.png)
Also, how big difference is expected in pixel values between target and source for each color chip? In brown for example, I see the difference around 22 (102 vs 125) between target and source.

Best,
Ali

**Comment by alihamidisepehr** [comment]
Thank you @nfahlgren and @JorgeGtz for continuous improvement. You have a very progressive team and I'm thankful. That issue was fixed. 
One challenge I faced was that I had many images in which the colorchecker was moving or tilted. I couldn't find a function on PlantCV that automatically detects the colorchecker. I used `create_color_card_mask` but since the colorchecker has been moving, it takes some manual checking to make sure source mask is in the right place. I'm using x-rite passport 2 colorchecker. 
I was wondering if you have any plans for automatic colorchecker detection using OpenCV features on this regard like following to make the process a little more automatic:
https://docs.opencv.org/4.5.3/de/df4/group__color__correction.html
https://docs.opencv.org/4.5.3/d9/d53/classcv_1_1mcc_1_1CCheckerDetector.html
https://docs.opencv.org/4.5.3/db/d3e/classcv_1_1mcc_1_1CChecker.html#details

**Comment by nfahlgren** [comment]
Thanks @alihamidisepehr! Did you try `pcv.transform.find_color_card`? https://plantcv.readthedocs.io/en/stable/find_color_card/. It is sensitive to rotation so might not work. I had not seen that this was added to OpenCV, we will definitely look into it. Looks like it was added in OpenCV4, which we do not support yet but will in PlantCV4

**Comment by maxjfeldman** [comment]
I've experienced some of these same difficulties too. 

I've attached 2 functions I wrote to check card orientation (grayscale chips on top vs. bottom; or left vs. right) and rotate card so it is perpendicular. 

Code isn't really too well annotated but the fxns are pretty small. Maybe it is somewhere to start.


[color_card_rotation_fxns.txt](https://github.com/danforthcenter/plantcv/files/7500729/color_card_rotation_fxns.txt)
